examples/nlp_seq/models.py

```diff
@@ -90,146 +69,150 @@ def init(key, shape, dtype=np.float32):
+          None, pos_emb_shape, None)
+    else:
+      pos_embedding = self.param('pos_embedding',
+                                 cfg.posemb_init,
+                                 pos_emb_shape)
+    pe = pos_embedding[:, :length, :]
+
+    if inputs_positions is None:
+      # normal unpacked case:
+      return inputs + pe
+    else:
+      # for packed data we need to use known position indices:
+      return inputs + jnp.take(pe[0], inputs_positions, axis=0)
 
 
 class MlpBlock(nn.Module):
-  """Transformer MLP block."""
+  """Transformer MLP / feed-forward block.
 
-  def apply(self,
-            inputs,
-            mlp_dim,
-            out_dim=None,
-            dropout_rate=0.1,
-            deterministic=False,
-            kernel_init=nn.initializers.xavier_uniform(),
-            bias_init=nn.initializers.normal(stddev=1e-6)):
+  Args:
+    config: TransformerConfig dataclass containing hyperparameters.
+    out_dim: optionally specify out dimension.
+  """
+  config: TransformerConfig
+  out_dim: Optional[int] = None
+
+  @nn.compact
+  def __call__(self, inputs, deterministic=True):
     """Applies Transformer MlpBlock module."""
-    actual_out_dim = inputs.shape[-1] if out_dim is None else out_dim
-    x = nn.Dense(inputs, mlp_dim, kernel_init=kernel_init, bias_init=bias_init)
-    x = nn.gelu(x)
-    x = nn.dropout(x, rate=dropout_rate, deterministic=deterministic)
-    output = nn.Dense(
-        x, actual_out_dim, kernel_init=kernel_init, bias_init=bias_init)
-    output = nn.dropout(output, rate=dropout_rate, deterministic=deterministic)
+    cfg = self.config
+    actual_out_dim = (inputs.shape[-1] if self.out_dim is None
+                      else self.out_dim)
+    x = nn.Dense(cfg.mlp_dim,
+                 dtype=cfg.dtype,
+                 kernel_init=cfg.kernel_init,
+                 bias_init=cfg.bias_init)(inputs)
+    x = nn.relu(x)
+    x = nn.Dropout(rate=cfg.dropout_rate)(x, deterministic=deterministic)
+    output = nn.Dense(actual_out_dim,
+                      dtype=cfg.dtype,
+                      kernel_init=cfg.kernel_init,
+                      bias_init=cfg.bias_init)(x)
+    output = nn.Dropout(rate=cfg.dropout_rate)(
+        output, deterministic=deterministic)
     return output
 
 
-class Transformer1DBlock(nn.Module):
-  """Transformer layer (https://openreview.net/forum?id=H1e5GJBtDr)."""
+class Encoder1DBlock(nn.Module):
+  """Transformer decoder layer.
 
-  def apply(self,
-            inputs,
-            qkv_dim,
-            mlp_dim,
-            num_heads,
-            causal_mask=False,
-            padding_mask=None,
-            dropout_rate=0.1,
-            attention_dropout_rate=0.1,
-            deterministic=False):
-    """Applies Transformer1DBlock module.
+  Args:
+    config: TransformerConfig dataclass containing hyperparameters.
+  """
+  config: TransformerConfig
+
+  @nn.compact
+  def __call__(self,
+               inputs,
+               encoder_mask=None,
```

Comment:
can we make all arguments to `__call__()` kw-only?

See also https://github.com/google/flax/issues/231#issuecomment-716568710

Reply:
done.